hail/src/main/scala/is/hail/expr/ir/InferPType.scala

```diff
@@ -28,45 +48,102 @@ object InferPType {
         while(vit.hasNext) {
           val value = vit.next()
           InferPType(value, env)
-          assert(head.pType2 == value.pType2)
+          assert(head.pType2 isOfType value.pType2)
         }
 
         head.pType2
       }
       case Ref(name, _) => env.lookup(name)
       case In(_, t) => PType.canonical(t)
-      case MakeArray(_, t) => PType.canonical(t)
-      case MakeStream(_, t) => PType.canonical(t)
+        // TODO: not sure if preference is to continue using canonical here
+        // I could imagine a collection of non-canonical ptypes
+      case MakeArray(irs, _) => {
+        val it = irs.iterator
+        val head = it.next()
+        InferPType(head, env)
+        var allRequired = head.pType2.required
+
+        while(it.hasNext) {
+          val irElem = it.next()
+
+          InferPType(irElem, env)
+          assert(head.pType2 isOfType irElem.pType2)
+
+          if(allRequired == true && irElem.pType2.required == false) {
+            allRequired = false
+          }
+        }
+
+        PArray(head.pType2, allRequired)
+      }
+      case MakeStream(irs, _) => {
+        val allRequired = irs.forall(ir => {
+          InferPType(ir, env)
+
+          ir.pType2.required
+        })
+
+        irs(0).pType2.setRequired(allRequired)
+      }
       case MakeNDArray(data, shape, _) => {
         InferPType(data, env)
         InferPType(shape, env)
 
         val nElem = shape.pType2.asInstanceOf[PTuple].size
 
-        PNDArray(coerce[PArray](data.pType2).elementType, nElem)
+        PNDArray(coerce[PArray](data.pType2).elementType, nElem, data.pType2.required)
+      }
+      case ArrayRange(start: IR, stop: IR, step: IR) => {
+        InferPType(start, env)
+        InferPType(stop, env)
+        InferPType(step, env)
+
+        assert(start.pType2 isOfType stop.pType2)
+        assert(start.pType2 isOfType step.pType2)
+
+        val allRequired = start.pType2.required &&
+          stop.pType2.required &&
+          step.pType2.required
+
+        PArray(start.pType2, allRequired)
```

Comment:
The elements generated by a range are always present, so should be a required type.

Reply:
Hm, I asked this question, and Tim stated that if all 3 were missing this the range could be missing.

Reply:
Yep---but the elements of the range themselves can never be missing, even if the entire array is.

Reply:
(suggesting that it should be `PArray(PInt32(true), allRequired)`, not `PArray(PInt32(), true)`)

Reply:
Here is all 3 elements of the range (start, stop, stride) are required, the ArrayRange is required, so it would always be required.

Reply:
arcturus is right, here -- if any of the args is missing, you end up with a missing array. if all the args are defined, you end up with an array with no missing elements.

Reply:
> so it would always be required if PArray(PInt32(true), allRequired)

This means that the entire array can be missing if not all the args are required, but *given that the array is defined* (whether statically or at runtime), the elements cannot be missing.

Reply:
And if it isn't defined, its elements are not required, correct?

Reply:
No, the elements are still required. If the entire array is missing, there are no elements---and hence none of the elements in that case are ever missing.

Reply:
> And if it isn't defined, its elements are not required, correct?

I don't think this is a meaningful question. If the array isn't defined, what does it mean for its elements to be defined?

`PArray(PInt32(false), false))` would have one bit to track whether the entire array is missing, plus a bit-array tracking the missingness of the individual elements.
* If the array missing-bit is set, we will never read the rest, so it doesn't matter if the extra bit-array is there or not.
* If the array missing-bit is not set, then because the array was constructed by `Range` we know all the elements are non-missing, so the extra bit-array is always all zeroes.

Since the extra bit-array is never providing meaningful information, it's better to omit it, which is the type `PArray(PInt32(true), false)`.

Reply:
Why would we set the missingness bit at the element level? 

I thought the point of the missingness header was to encode the indices that are missing? Why would we do this for a PInt32?

Reply:
Let's be careful to distinguish missingness bits (which exist only at run-time), and requiredness bits (which exist only at compile time).

The requiredness bit in `PArray(PInt32(false))` is recording that the ints inside the array may be missing, so at run-time we will have to allocate an array of missingness bits to track which elements are missing.

On the other hand, the requiredness bit in `PArray(PInt32), false)` is recording that the array as a whole may be missing. What that means in terms of data layout depends on if this `PArray` type is nested inside some other collection type.

Again, it might help you (it does for me) to logically identify `PInt32(false)` with something like `Option[Int]`. The fact that `PArray` can pack the missingness bits for all its children into a dense array is just an optimization. A free-standing `PInt32(false)` is still a meaningful thing.

Reply:
This clears it up for me, thanks.

So the required bit is always referencing whether the type is provably (at compile time) non-null.

A PArray with elementType.required == false will have a missing header code-generated, while a PArray with required == true will always initialized to an array (even with length 0, []).

Namely we cannot infer the requiredeness of the PArray that is the physical type of the MakeArray node, from the MakeArray node's elements (those will only affect the PArray's elementType). So for now, we trivially set all collections to be required = true.

Reply:
Ah, but that isn't the same thing! Arcturus recommended (PArray(PInt32(true), allRequired)) which is the opposite of that.

Reply:
Arcturus is right. We don't want to set all collections to be required = true, and this is a good example why.

If we execute the IR `ArrayRange(start, stop, step)`, and at run-time one of the inputs happens to be missing, what should we do? We've decided the right answer (almost always) is that if any inputs to an operation are missing, the result should be missing. In this case, it's the entire array that is missing.

But no matter what, if we are able to construct a non-missing array, we know statically that it will not have any missing elements, so the value type should be required.

`MakeArray`, on the other hand, should always be a `PArray(required=true)`. No matter what the arguments are, we are constructing an array, so we know it can't ever be missing.

Reply:
Makes sense.